### DISCUSSION TITLE
Secret with special/escape characters fails when used

### DISCUSSION THREAD

**Discussion by derek-ef** [community/community#25541] [Actions]
<p>I am using a macos-latest executor to build a React Native app for iOS. The build itself works fine but deploying to TestFlight (via Fastlane) fails due to incorrect formatting of the Apple ID session token. I am generating the token locally and then saving it as a secret in the GitHub repo per this <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets" rel="nofollow noopener">documentation</a>. The Fastlane script expects the token to be available via environment variable FASTLANE_SESSION so I am exporting the value of the secret as the environment variable per the GitHub actions documentation. It should be noted that the token string has special and escaped characters and Apple expects it to be formatted exactly as it is generated. I have tested locally and the session tokens I have generated are valid - they only fail when being read from the secret store in GitHub Actions. The Fastlane step fails with the following error:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/1/1c151033130444f194a58d73ad67eef18288dfcd.png" data-download-href="/uploads/short-url/40quOCEahIuFDGg4HzcJxTAu1Q1.png?dl=1" title="Screen Shot 2019-12-03 at 10.40.36 AM.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/47192444/181091763-c2e02a92-bf28-4337-97a6-343c37b8b25b.png" alt="Screen Shot 2019-12-03 at 10.40.36 AM.png" data-base62-sha1="40quOCEahIuFDGg4HzcJxTAu1Q1" width="690" height="88" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/1/1c151033130444f194a58d73ad67eef18288dfcd_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-12-03 at 10.40.36 AM.png</span><span class="informations">788×101 18.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Normal login (username and password) fails because 2FA is required - hence the use of the session token.</p>
<p>Relevant Actions yaml:</p>
<pre><code>deply-ios-qa:
    runs-on: macos-latest
    env:
      MATCH_PASSWORD: ${{ secrets.MATCH_PASSWORD }}
      FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD: $${{ secrets.FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD }}
      FASTLANE_PASSWORD: $${{ secrets.FASTLANE_PASSWORD }}
      FASTLANE_SESSION: $${{ secrets.FASTLANE_SESSION }}
    steps:
      - name: Checkout
        uses: actions/checkout@v1

      - name: Install gems
        run: |
          bundle install

      - name: Build iOS app
        run: bundle exec fastlane ios deploy_testflight_qa
</code></pre>
<p>Is there a way to confirm that the secret is coming out in the incorrect format? Is GitHub actions doing anything during the saving/reading of the secret that could affect the formatting? Again, Apple expects this string to be formatted exactly as their API gives it to you including spaces, special characters and escaped characters. Also of note: an identical Fastlane script with a similarly accessed secret works on another project in CircleCI so I know the workflow should succeed if the session token is correct.</p>

**Comment by thboop** [comment]
<p>Hey <a class="mention" href="/derek-ef">@derek-ef</a> ,</p>
<p>Are you intending to prepend the <code>FASTLANE_PASSWORD</code> and <code>FASTLANE_SESSION</code> with a <code>$</code>, </p>
<pre><code>FASTLANE_PASSWORD: ${{ secrets.FASTLANE_PASSWORD }}
      FASTLANE_SESSION: ${{ secrets.FASTLANE_SESSION }}
</code></pre>
<p>Would be the way to set those environmental variables to the value of those secrets.</p>
<p>Edit:  FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD looks like it is getting prepended as well.</p>

**Comment by derek-ef** [comment, marked as answer]
<p>Well, I suppose the string would be different if it had a $ at the beginning… yesterday was a long day.</p>
<p>Thanks for your response <a class="mention" href="/thboop">@thboop</a>, your suggestion worked with one addition. I had to add single quotes around the string to make sure the format was kept exactly so my env var looks like this:</p>
<pre><code>FASTLANE_SESSION: '${{ secrets.FASTLANE_SESSION }}'
</code></pre>
<p>On that topic, is there a good way to debug something like this? Obviously printing secrets to the console is masked in Actions which makes a lot of sense but having a way to run the action locally in CLI with local secrets defined would be a great improvement to developer experience. Is anything like this on the roadmap for Actions?</p>
